File: backend/src/main/java/com/fptgang/backend/model/Project.java
package com.fptgang.backend.model;

import com.fptgang.backend.util.Searchable;
import jakarta.annotation.Nullable;
import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;
import org.jetbrains.annotations.NotNull;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

@Entity
@Table(name = "projects")
@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Project {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long projectId;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "project_category_id", nullable = false)
    private ProjectCategory category;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "client_id", nullable = false)
    private Account client;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "staff_id")
    private Account staff;

    @Column(columnDefinition = "NVARCHAR(255)", length = 255, nullable = false)
    @Searchable
    private String title;

    @Column(columnDefinition = "TEXT", length = 100_000, nullable = false)
    @Searchable
    private String description;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private ProjectStatus status;

    @Column(nullable = false)
    private LocalDateTime startDate;

    @Column(nullable = false, precision = 10, scale = 2)
    private BigDecimal minBudget;

    @Column(nullable = false, precision = 10, scale = 2)
    private BigDecimal maxBudget;

    @CreationTimestamp
    private LocalDateTime createdAt;

    @UpdateTimestamp
    private LocalDateTime updatedAt;

    @Column(nullable = false, columnDefinition = "BOOLEAN DEFAULT TRUE")
    @Builder.Default
    private Boolean isVisible = true;

    @OneToMany(mappedBy = "project", fetch = FetchType.LAZY)
    @Builder.Default
    @ToString.Exclude
    @EqualsAndHashCode.Exclude
    private List<File> files = new ArrayList<>();

    @OneToMany(mappedBy = "project", fetch = FetchType.LAZY)
    @Builder.Default
    @ToString.Exclude
    @EqualsAndHashCode.Exclude
    @OrderBy("proposalId ASC")
    private List<Proposal> proposals = new ArrayList<>();

    @OneToOne(mappedBy = "project", fetch = FetchType.EAGER)
    @Nullable
    private Contract contract;

    // Cascading milestones with project
    @OneToMany(mappedBy = "project", cascade = CascadeType.ALL, fetch = FetchType.LAZY)
    @Builder.Default
    @ToString.Exclude
    @EqualsAndHashCode.Exclude
    @OrderBy("milestoneId ASC")
    private List<Milestone> milestones = new ArrayList<>();

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "active_milestone_id")
    @ToString.Exclude
    @EqualsAndHashCode.Exclude
    @Nullable
    private Milestone activeMilestone;

    // Cascading project skills with project
    @OneToMany(mappedBy = "project", cascade = CascadeType.ALL, fetch = FetchType.LAZY, orphanRemoval = true)
    @Builder.Default
    private List<ProjectSkill> requiredSkills = new ArrayList<>();

    @OneToMany(mappedBy = "project", fetch = FetchType.LAZY)
    @Builder.Default
    @ToString.Exclude
    @EqualsAndHashCode.Exclude
    @OrderBy("messageId ASC")
    private List<Message> messages = new ArrayList<>();

    @OneToMany(mappedBy = "project", fetch = FetchType.LAZY)
    @Builder.Default
    @ToString.Exclude
    @EqualsAndHashCode.Exclude
    @OrderBy("reportId ASC")
    private List<Report> reports = new ArrayList<>();

    public enum ProjectStatus {
        OPEN,
        PAUSED,
        IN_PROGRESS,
        TERMINATED,
        FINISHED
    }

    @Enumerated(EnumType.STRING)
    @Nullable
    private TerminationReason terminationReason;

    @Column(nullable = false, columnDefinition = "BOOLEAN DEFAULT FALSE")
    @Builder.Default
    private Boolean toTerminate = false;

    public enum TerminationReason {
        /**
            Other reasons before the contract is made
            e.g. Client didn’t accept any proposal by startDate
                 No proposals were submitted by startDate
         */
        OTHER,

        /** Contract remained unsigned by the start date */
        CONTRACT_UNSIGNED,

        /** Client can request termination before 2 days past the current milestone deadline
            Become effective starting from the next milestone  */
        CLIENT_REQUEST_TERMINATION,

        /** Project terminated by staff after conflict resolution */
        STAFF_DECISION
    }

    @Nullable
    public Account getFreelancer() {
        return getContract() == null ? null : getContract().getFreelancer();
    }
}

File: gendb/src/engine/model/Project.ts
import {escapeSingleQuotes} from "../utils.js";
import {Account} from "./Account.js";
import {Milestone} from "./Milestone.js";
import {MilestoneStatus} from "./MilestoneStatus.js";
import {ProjectRequiredSkill} from "./ProjectRequiredSkill.js";
import {ProjectStatus} from "./ProjectStatus.js";
import Contract from "./Contract";

export enum TerminationReason {
  /**
   Other reasons before the contract is made
   e.g. Client didn’t accept any proposal by startDate
   No proposals were submitted by startDate
   */
  OTHER = "OTHER",

  /** Contract remained unsigned by the start date */
  CONTRACT_UNSIGNED = "CONTRACT_UNSIGNED",

  /** Client failed to fund milestone within 24 hours */
  CLIENT_NO_FUNDS = "CLIENT_NO_FUNDS",

  /** Client can request termination before 2 days past the current milestone deadline
   Become effective starting from the next milestone  */
  CLIENT_REQUEST_TERMINATION = "CLIENT_REQUEST_TERMINATION",

  /** Freelancer abandoned work for 2 days past the milestone deadline */
  FREELANCER_ABANDONED_WORK = "FREELANCER_ABANDONED_WORK",

  /** Project terminated by staff after conflict resolution */
  STAFF_INTERVENTION_DECISION = "STAFF_INTERVENTION_DECISION"
}

export class Project {
  project_id: number;
  created_at: Date | null;
  description: string;
  is_visible: boolean;
  max_budget: number;
  min_budget: number;
  start_date: Date;
  status: ProjectStatus;
  termination_reason: TerminationReason | null;
  title: string;
  to_terminate: boolean;
  updated_at: Date | null;
  project_category_id: number;
  client_id: number;
  staff_id: number | null;

  // not dumpable
  client: Account;
  project_skills: ProjectRequiredSkill[];
  milestones: Milestone[];
  activeMilestone: Milestone | null;
  contract: Contract | undefined;

  constructor(data: Partial<Project>) {
    this.project_id = data.project_id || 0;
    this.created_at = data.created_at ? new Date(data.created_at) : null;
    this.description = data.description || '';
    this.is_visible = data.is_visible ?? true;
    this.max_budget = data.max_budget ?? 0;
    this.min_budget = data.min_budget ?? 0;
    this.start_date = data.start_date ?? new Date();
    this.status = data.status || ProjectStatus.OPEN;
    this.to_terminate = data.to_terminate || false;
    this.termination_reason = data.termination_reason || null;
    this.title = data.title || '';
    this.updated_at = data.updated_at ? new Date(data.updated_at) : null;
    this.project_category_id = data.project_category_id || 0;
    this.client_id = data.client_id || 0;
    this.staff_id = data.staff_id || null;

    this.client = data.client || {} as Account;
    this.project_skills = data.project_skills || [];
    this.milestones = data.milestones || [];
    this.activeMilestone = data.activeMilestone || null;
  }

  static dump(projects: Project[]): string {
    if (projects.length === 0) return '';

    const values = projects.map(project => {
      return `(${[
        project.project_id,
        project.created_at ? `'${project.created_at.toISOString().slice(0, 19)}.000000'` : 'NULL',
        `'${escapeSingleQuotes(project.description)}'`,
        project.is_visible ? 1 : 0,
        project.max_budget.toFixed(2),
        project.min_budget.toFixed(2),
        `'${project.start_date.toISOString().slice(0, 19).replace('T', ' ')}.000000'`,
        `'${project.status}'`,
        project.termination_reason ? `'${escapeSingleQuotes(project.termination_reason)}'` : 'NULL',
        project.to_terminate ? 1 : 0,
        project.activeMilestone ? project.activeMilestone.milestoneId : 'NULL',
        `'${escapeSingleQuotes(project.title)}'`,
        project.updated_at ? `'${project.updated_at.toISOString().slice(0, 19)}.000000'` : 'NULL',
        project.project_category_id,
        project.client_id,
        project.staff_id ? project.staff_id : 'NULL'
      ].join(', ')})`;
    }).join(',\n');

    return `INSERT INTO \`projects\` (\`project_id\`, \`created_at\`,
                                      \`description\`,
                                      \`is_visible\`, \`max_budget\`,
                                      \`min_budget\`, \`start_date\`,
                                      \`status\`, \`termination_reason\`,
                                      \`to_terminate\`, \`active_milestone_id\`, \`title\`,
                                      \`updated_at\`, \`project_category_id\`,
                                      \`client_id\`, \`staff_id\`)
            VALUES ${values};`;
  }

  nextMilestone(): Milestone | undefined {
    return this.milestones.find(milestone => milestone.status === MilestoneStatus.PENDING);
  }
}